Nav Block: Show justification controls for vertical variant

mkaz opened the pull request:
## Description

- Enable justify content controls to show for both horiz and vert variants
- Filter out space-between from vertical controls
 
Closes #29459 

## How has this been tested?

1. Add Navigation block (vertical)
2. Confirm the left-center-right controls show and work as expected
3. Confirm the space-between controls do not show
4. Test Navigation block (horizontal) space-between shows and no regression

## Screenshots 

See comments below.

mkaz committed:
Show justification controls for vertical variant

- Don't limit controls to show only on horizontal
- This enables controls to show for both horiz and vert
- The vertical CSS specifies a block so will require tweaking
- Filter out space-between from vertical controls

jasmussen commented:
Thank you Marcus, it's already been a long day for me so I'm going to take a look at this one tomorrow!

jasmussen committed:
Fix aligments.

jasmussen commented:
Fixed the alignments. Let's get a sanity check and merge.

![now](https://user-images.githubusercontent.com/1204802/112975401-e0841e00-9153-11eb-8489-edd57490b5db.gif)

mkaz commented:
@jasmussen Ah, right `flex-direction: column` for vertical. I knew I was missing something straight forward.

Testing it out and it doesn't seem to be working properly for horizontal, the vertical list looks and works fine but horizontal changes are not being seen with the block selected, but only show when not selected. See GIF below,

![alignment-not-selected](https://user-images.githubusercontent.com/45363/113015892-cf76e500-9132-11eb-9925-2545f57ee86a.gif)

jasmussen commented:
Try deselecting the horizontal version once you apply an alignment. If that works, then it's a known issue partially mitigated by https://github.com/WordPress/gutenberg/pull/29771

Edit: ah I see you did try deselecting. Yeah that should be present also in trunk.

mkaz commented:
@jasmussen Confirmed, same behavior exists on trunk! No regression on this PR :heavy_check_mark:

mkaz commented:
I did more testing on other themes and in different regiions, both the Navigation Editor, and Site Editor Screens.

![vert-nav-site-editor](https://user-images.githubusercontent.com/45363/113074132-59e53600-917f-11eb-9cf3-fcb3fd22505c.gif)

![vert-nav-editor-screen](https://user-images.githubusercontent.com/45363/113074143-5f428080-917f-11eb-8900-394489edc05a.gif)

jasmussen commented:
Awesome testing. Actually the fact that it affects the navigation screen — it probably shouldn't, which is to say @talldan might want to know.

mkaz committed:
Add attribute so Nav Editor can hide justify

- Adds an additional attribute to showJustifyControls to navigation
block so the Nav Editor screen can hide the controls, since that screen
is intended to create/edit nav menus and not format/style.

- Use 'showJustifyControls' since the positive is easier to reason and
there is already a 'showSubMenuIcon' attribute that also defaults to
true

mkaz commented:
@jasmussen Good catch on the navigation editor screen! I tested it there and confirmed it worked, but didn't think if it makes sense there or not. Considering that screen is to create and edit menus for classic themes, the layout format won't work and should not be there.

In 9ebbbab, I added a new attribute that can be used to disable the controls, and updated the Nav Editor to pass in this attribute so the justify controls will not show. Tested and confirmed all are showing everywhere else, but not there. 👍 

@talldan Can you confirm this makes sense, and works as you would expect? 🙏

talldan commented:
@mkaz Thanks for the ping. @jasmussen is right in that it shouldn't be on the navigation screen.

This navigation screen previously disabled this using a prop that's passed into the block's edit function, but it looks like this PR removes that prop:
https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-navigation/src/filters/remove-edit-unsupported-features.js#L17

Instead of removing it, you should be able to default it to `true` so that it shows in both orientations, which the nav editor will then override.

I think it'd be good to avoid using an attribute for this purpose, as attributes shouldn't really be for editor side settings, more for the frontend representation of a block.

Hope that helps.

talldan commented:
BTW - there is an issue to discuss better ways of enabling/disabling features in the nav block (https://github.com/WordPress/gutenberg/issues/30007).

I think content justification is one of the easier ones, in that it could be a candidate for becoming a block 'supports' feature. I know there's something very similar on Buttons block as well. One thing to check is that the changes here don't diverge too much from the Buttons implementation.

mkaz commented:
@talldan After looking into it and rediscovering why we opted for this implementation in the first place. The way the justification works is it applies to the inner elements and can't be applied to the wrapping element the same for all uses. The supports mechanism doesn't allow for the level of customization needed.

An example can be seen for Social Links implementation here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/social-links/edit.js#L119

the Social Links uses its own function and specific CSS to apply properly to the correct inner elements.

talldan commented:
Ok. My comment here still applies - https://github.com/WordPress/gutenberg/pull/30351#issuecomment-814730380

jasmussen commented:
Alright, to follow up on this one, I rebased to fix the conflict, and I made an additional fix. I realized text in menu items should be right-aligned. Before:

<img width="744" alt="Screenshot 2021-04-13 at 12 23 27" src="https://user-images.githubusercontent.com/1204802/114538910-311b7100-9c54-11eb-891e-75488ff22b6c.png">

After:

<img width="737" alt="Screenshot 2021-04-13 at 12 32 27" src="https://user-images.githubusercontent.com/1204802/114539032-53ad8a00-9c54-11eb-8117-98f546370b07.png">

Note that I tested the navigation screen again, and I couldn't see the justification controls there:

<img width="554" alt="Screenshot 2021-04-13 at 12 25 45" src="https://user-images.githubusercontent.com/1204802/114539072-632cd300-9c54-11eb-8da1-143f8854095f.png">

So that issue appears fixed. Or was there anything else we needed to do?

jasmussen committed:
Right align text.

mkaz committed:
Use hasItemJustificationCOntrols prop instead of attribute

mkaz commented:
@talldan Ok, updated with I think what you were saying. I was confusing props and attributes before.

I went back to previous prop that was in place and switching that to default true, basically does everything we want, plus the additional updates in this PR for style and hiding space-between.

If you can take a look now, I think its right and ready, and sorry for any confusion. 👍

talldan reviewed (approved):
Thanks for making those changes. 🎉 

The only issue I'm seeing when testing is that the spacer block is very narrow for vertical menus, but it doesn't work all that well in `trunk` either (it is wider, but seems to always show `0px` when resizing.)

jasmussen commented:
> The only issue I'm seeing when testing is that the spacer block is very narrow for vertical menus, but it doesn't work all that well in trunk either (it is wider, but seems to always show 0px when resizing.)

Let me investigate that before I merge.

Thank you for the review!

jasmussen commented:
In my theme I'm seeing the same before:

<img width="848" alt="Screenshot 2021-04-14 at 09 37 24" src="https://user-images.githubusercontent.com/1204802/114672420-6b911680-9d05-11eb-8ea2-e5f5a03c1688.png">

<img width="795" alt="Screenshot 2021-04-14 at 09 38 27" src="https://user-images.githubusercontent.com/1204802/114672424-6cc24380-9d05-11eb-8136-141c2c59dad8.png">

and after:

<img width="746" alt="Screenshot 2021-04-14 at 09 39 54" src="https://user-images.githubusercontent.com/1204802/114672437-6fbd3400-9d05-11eb-94be-46f8c9c7172d.png">
<img width="753" alt="Screenshot 2021-04-14 at 09 40 02" src="https://user-images.githubusercontent.com/1204802/114672441-70ee6100-9d05-11eb-9e6f-85d2038bfa72.png">

If you have a screenshot or steps to reproduce, I'm happy to take a look.

talldan commented:
@jasmussen Hmm, not much in the way of repro steps, just added the spacer to a vertical nav block with this branch checked out. Seems the same for Twenty Twenty and Twenty Twenty One (and also the same with theme styles disabled in the editor):
<img width="261" alt="Screenshot 2021-04-14 at 4 52 49 pm" src="https://user-images.githubusercontent.com/677833/114682909-1f17fc00-9d42-11eb-8f81-89e0c7720587.png">

<img width="643" alt="Screenshot 2021-04-14 at 4 53 46 pm" src="https://user-images.githubusercontent.com/677833/114682928-23dcb000-9d42-11eb-81dc-4562b8812822.png">

I also notice if you switch between horizontal and vertical nav menus, the spacer seems to retain the width from when it was horizontal. And some `NaN` errors in the console.

Just to check if your styles have rebuilt properly, is your `wp-block-navigation__container` set to `display: flex` when you inspect it? That seems to be the change that made this happen. Switching back to `block` makes the spacer full width again.

jasmussen commented:
Okay I was able to reproduce! Thank you. Marcus don't merge this until I have a chance to look at it :D

jasmussen committed:
Restore property.

jasmussen commented:
Okay I restored a property which makes it work for me:

<img width="787" alt="Screenshot 2021-04-14 at 13 28 05" src="https://user-images.githubusercontent.com/1204802/114703310-65129700-9d25-11eb-9f1d-6be7c69df3e9.png">

Marcus please test that alignments left right and center as well as the spacer in the vertical nav block work for you also.

mkaz commented:
This all tested well for me 👍 
Thanks all for the collab and teamwork! 🎉
